### ISSUE TITLE
Add generic object/artifact reference attestation field type

### ISSUE THREAD

**Issue by marcelamelara** [in-toto/attestation#117]
For attestation predicates that reference other attestations or artifacts, such as the `evidence` fields in SCAI (#112) or SLSA v1.0, it's beneficial for scalability to provide a reference to the object in question, rather than including the full object. The SCAI specification proposes an Object Reference format (see [SCAI spec doc](https://arxiv.org/abs/2210.05813), Sec. 3), which can serve as the foundation for a generic object reference type:

```
{ 
    "name": "<NAME>",
     "digest": { "<ALGORITHM>": "<HEX VALUE>", ... },
     "locationURI": "<URI>",
     "objectType": "<OBJECT TYPE>" // optional
}
```

This object follows a common pattern as seen in the (extended) attestation `subject` field (#5) or the artifact reference in SLSA v1.0 [Artifact Reference](https://github.com/slsa-framework/slsa/blob/e60f7426f501f34e8b41e9b4da594bb7ec1c57bb/docs/provenance/v1.0.cue#L37), so there is precedent to formalize such a field type in attestations.

Examples for how this field type might be used can be found as part of the [SCAI predicate spec](https://github.com/in-toto/attestation/blob/main/spec/predicates/scai.md#examples)

**Comment by marcelamelara** [comment]
Nice! I like the idea of having the MIME type in there, instead of the less descriptive `objectType` I have in there right now. 

I hadn't considered `References` as predicates, but SBOM seems like a great use case for this. I had been imagining that even fields like `subject` may become of type `Reference`, but this is something we could discuss.

**Comment by TomHennen** [comment]
Oh I guess that does seem like a different use case.  For our purposes we just need to communicate that this document exists and don't necessarily need to reference it from within some pre-existing in-toto statement.

**Comment by marcelamelara** [comment]
So, in essence, the in-toto attestation then really just becomes a signature over the document without having to sign the entire document directly. I would expect this to be a pretty common use case in practice, so I think we definitely want to support this.

**Comment by marcelamelara** [comment]
This issue may be relevant to #27 and #2 as well.

**Comment by TomHennen** [comment]
Yes I think so, but I wonder what the location field would be if the attestation is contained within the bundle but not necessarily at any specific URI?  Just unset?  Some special value like "<in-bundle>"?

Also, revisiting this as a predicate, @AdamZWu suggested that perhaps we could use the same object both as predicate and as type that can be used more generally.  I think the only hiccup would be that referrerId probably wouldn't make sense if it's not in a predicate (which I think might be a reason to add an 'attestorId' to the statement layer (e.g. then referrerId is redundant and can be removed).  WDYT?

**Comment by marcelamelara** [comment]
> I wonder what the location field would be if the attestation is contained within the bundle but not necessarily at any specific URI? Just unset? Some special value like ""?

Right, this is a good question. In the original SCAI spec document, the `locationURI` field is a required field, but I think selecting a special value for when the location is "local" would make sense to avoid/reduce ambiguity.

> we could use the same object both as predicate and as type that can be used more generally. I think the only hiccup would be that referrerId probably wouldn't make sense if it's not in a predicate

What's the reasoning behind including the `referrerId` in the predicate? Is this to be able to link the reference back to the signer/creator of the SBOM? Or put differently, are you assuming that Statements might be distributed without the DSSE Envelope containing the signer's key?

**Comment by TomHennen** [comment]
Sorry for the late reply, it's been a busy few weeks.

> What's the reasoning behind including the referrerId in the predicate? Is this to be able to link the reference back to the signer/creator of the SBOM? Or put differently, are you assuming that Statements might be distributed without the DSSE Envelope containing the signer's key?

It's similar to why SLSA provenance has the builder id, SCAI has an ID, etc...   It gives some more user readable way to indicate who referred to this thing that is not tied directly to the specific key material (which might change!).

**Comment by marcelamelara** [comment]
Addressed in #143.
